pkgs/applications/networking/mailreaders/notmuch/default.nix

```diff
@@ -133,7 +133,28 @@ stdenv.mkDerivation rec {
   + lib.optionalString withSfsexp ''
     cp notmuch-git $out/bin/notmuch-git
     wrapProgram $out/bin/notmuch-git --prefix PATH : $out/bin:${lib.getBin git}/bin
-  '';
+  '' + lib.optionalString withVim ''
+    make -C vim DESTDIR="$out/share/vim-plugins/notmuch" prefix="" install
+    mkdir -p $out/share/nvim
+    ln -s $out/share/vim-plugins/notmuch $out/share/nvim/site
+  '' + lib.optionalString (withVim && withRuby) (let
+    gemEnv = buildEnv {
```

Comment:
Interesting! Now I'm surprised how `ruby.withPackages` doesn't fit here, because it doesn't add `"/lib"` to `pathsToLink` (found this out locally).

In anycase, it'd still be great if you'd transition into `finalAttrs`, as it would allow someone to override this derivation if they need from some reason more gems installed.

Here are changes I tried, based upon yours:

```diff
diff --git i/pkgs/applications/networking/mailreaders/notmuch/default.nix w/pkgs/applications/networking/mailreaders/notmuch/default.nix
index c9b0e158acf2..8348919d1769 100644
--- i/pkgs/applications/networking/mailreaders/notmuch/default.nix
+++ w/pkgs/applications/networking/mailreaders/notmuch/default.nix
@@ -15,12 +15,12 @@
 , withVim ? true
 }:
 
-stdenv.mkDerivation rec {
+stdenv.mkDerivation (finalAttrs: {
   pname = "notmuch";
   version = "0.38.3";
 
   src = fetchurl {
-    url = "https://notmuchmail.org/releases/notmuch-${version}.tar.xz";
+    url = "https://notmuchmail.org/releases/notmuch-${finalAttrs.version}.tar.xz";
     hash = "sha256-mvRsyA2li0MByiuu/MJaQNES0DFVB+YywPPw8IMo0FQ=";
   };
 
@@ -137,29 +137,27 @@ stdenv.mkDerivation rec {
     make -C vim DESTDIR="$out/share/vim-plugins/notmuch" prefix="" install
     mkdir -p $out/share/nvim
     ln -s $out/share/vim-plugins/notmuch $out/share/nvim/site
-  '' + lib.optionalString (withVim && withRuby) (let
-    gemEnv = buildEnv {
-      name = "notmuch-vim-gems";
-      paths = with ruby.gems; [ mail ];
-      pathsToLink = [ "/lib" "/nix-support" ];
-    };
-  in ''
+  '' + lib.optionalString (withVim && withRuby) ''
     PLUG=$out/share/vim-plugins/notmuch/plugin/notmuch.vim
     cat >> $PLUG << EOF
-      let \$GEM_PATH=\$GEM_PATH . ":${gemEnv}/${ruby.gemPath}"
+      let \$GEM_PATH=\$GEM_PATH . ":${finalAttrs.passthru.gemEnv}/${ruby.gemPath}"
       let \$RUBYLIB=\$RUBYLIB . ":$out/${ruby.libPath}/${ruby.system}"
       if has('nvim')
     EOF
-    for gem in ${gemEnv}/${ruby.gemPath}/gems/*/lib; do
+    for gem in ${finalAttrs.passthru.gemEnv}/${ruby.gemPath}/gems/*/lib; do
       echo "ruby \$LOAD_PATH.unshift('$gem')" >> $PLUG
     done
     echo 'endif' >> $PLUG
-  '');
+  '';
 
   passthru = {
-    pythonSourceRoot = "notmuch-${version}/bindings/python";
+    pythonSourceRoot = "notmuch-${finalAttrs.version}/bindings/python";
     tests.version = testers.testVersion { package = notmuch; };
-    inherit version;
+    gemEnv = buildEnv {
+      name = "notmuch-vim-gems";
+      paths = with ruby.gems; [ mail ];
+      pathsToLink = [ "/lib" "/nix-support" ];
+    };
 
     updateScript = gitUpdater {
       url = "https://git.notmuchmail.org/git/notmuch";
@@ -176,4 +174,4 @@ stdenv.mkDerivation rec {
     platforms   = platforms.unix;
     mainProgram = "notmuch";
   };
-}
+})
```

And here is how I tested the overriding mechanism:

Without overriding:

```
$ nix show-derivation --impure --expr 'with (builtins.getFlake "'$PWD'").legacyPackages.${builtins.currentSystem}; notmuch.overrideAttrs(old: {
})' | jq --raw-output '.[].env.postInstall'
moveToOutput bin/notmuch-emacs-mua $emacs
make -C bindings/ruby install \
  vendordir=$out/lib/ruby \
  SHELL=$SHELL \
  $makeFlags "${makeFlagsArray[@]}" \
  $installFlags "${installFlagsArray[@]}"
cp notmuch-git $out/bin/notmuch-git
wrapProgram $out/bin/notmuch-git --prefix PATH : $out/bin:/nix/store/apl2ibxqrxbbj6z79nfgv892999z1zyv-git-2.44.0/bin
make -C vim DESTDIR="$out/share/vim-plugins/notmuch" prefix="" install
mkdir -p $out/share/nvim
ln -s $out/share/vim-plugins/notmuch $out/share/nvim/site
PLUG=$out/share/vim-plugins/notmuch/plugin/notmuch.vim
cat >> $PLUG << EOF
  let \$GEM_PATH=\$GEM_PATH . ":/nix/store/a4s4ymn7k90yh9bld5zic587n1kg0k1v-notmuch-vim-gems/lib/ruby/gems/3.1.0"
  let \$RUBYLIB=\$RUBYLIB . ":$out/lib/ruby/3.1.0/x86_64-linux"
  if has('nvim')
EOF
for gem in /nix/store/a4s4ymn7k90yh9bld5zic587n1kg0k1v-notmuch-vim-gems/lib/ruby/gems/3.1.0/gems/*/lib; do
  echo "ruby \$LOAD_PATH.unshift('$gem')" >> $PLUG
done
echo 'endif' >> $PLUG
```

With overriding:

```
$ nix show-derivation --impure --expr 'with (builtins.getFlake "'$PWD'").legacyPackages.${builtins.currentSystem}; notmuch.overrideAttrs(old: {
  passthru = old.passthru // {
    gemEnv = buildEnv {
      name = "notmuch-more-gems";
      paths = with ruby.gems; [
        mail
        clamp
      ];
    };
  };
})' | jq --raw-output '.[].env.postInstall'
moveToOutput bin/notmuch-emacs-mua $emacs
make -C bindings/ruby install \
  vendordir=$out/lib/ruby \
  SHELL=$SHELL \
  $makeFlags "${makeFlagsArray[@]}" \
  $installFlags "${installFlagsArray[@]}"
cp notmuch-git $out/bin/notmuch-git
wrapProgram $out/bin/notmuch-git --prefix PATH : $out/bin:/nix/store/apl2ibxqrxbbj6z79nfgv892999z1zyv-git-2.44.0/bin
make -C vim DESTDIR="$out/share/vim-plugins/notmuch" prefix="" install
mkdir -p $out/share/nvim
ln -s $out/share/vim-plugins/notmuch $out/share/nvim/site
PLUG=$out/share/vim-plugins/notmuch/plugin/notmuch.vim
cat >> $PLUG << EOF
  let \$GEM_PATH=\$GEM_PATH . ":/nix/store/hadibs48n5cympz47k3ypag89yvbd6x1-notmuch-more-gems/lib/ruby/gems/3.1.0"
  let \$RUBYLIB=\$RUBYLIB . ":$out/lib/ruby/3.1.0/x86_64-linux"
  if has('nvim')
EOF
for gem in /nix/store/hadibs48n5cympz47k3ypag89yvbd6x1-notmuch-more-gems/lib/ruby/gems/3.1.0/gems/*/lib; do
  echo "ruby \$LOAD_PATH.unshift('$gem')" >> $PLUG
done
echo 'endif' >> $PLUG
```

Note how the path to the `gemEnv` changed - as requested. It won't be possible to perform such a change so elegantly without `finalAttrs`.

Reply:
Okay, makes sense!

Reply:
But @RyanGibb you didn't define `gemEnv` in `passthru`, and then used `finalAttrs.passthru`...